rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,200 @@
+---
+feature: nixpkgs_version_normalization
+start-date: 2021-09-24
+author: Anderson Torres
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+---
+
+# Summary
+[summary]: #summary
+
+Normalize the `version` attribute used in Nixpkgs' expresions.
+
+# Motivation
+[motivation]: #motivation
+
+Nowadays, the most commonly used format for the `pname` and `version` attributes
+along the Nixpkgs' expressions is:
+
+- For stable releases:
+  - `pname` reflects the software name;
+  - `version` reflects the released version.
+- For unstable releases:
+  - `pname` reflects the software name;
+  - `version` is a string following the format `unstable-YYYY-MM-DD`, where
+    `YYYY-MM-DD` denotes the date when the code was generated.
+
+This is a simple and easy-to-understand format.
+
+However, it does not map very well with the Nix function
+`builtins.parseDrvName`:
+
+```example
+# expected: { name = "mpv"; version = "0.35.15"; }
+nix-repl> builtins.parseDrvName "mpv-0.35.15"
+builtins.parseDrvName "mpv-0.35.15"
+{ name = "mpv"; version = "0.35.15"; }
+
+# expected: { name = "mpv"; version = "unstable-2021-05-03"; }
+nix-repl> builtins.parseDrvName "mpv-unstable-2021-05-03"
+builtins.parseDrvName "mpv-unstable-2021-05-03"
+{ name = "mpv-unstable"; version = "2021-05-03"; }
+```
+
+It happens because the `version` attribute returned by `builtins.parseDrvName`
+starts with a digit. This is not a bug, but a deliberate design decision of Nix
+specification; therefore we should strive to follow it, neither circumventing
+nor ignoring it.
+
+That being said, `version` attribute should start with a digit.
+
+Furthermore, `version` attribute should follow the semantics for upgrading as
+stated in `nix-env` manpage.
```

Comment:
I find this motivation a bit lacking. From my experience there is quite strong consensus that `nix-env` and `builtins.parseDrvName` have some inherent flaws (which have already caused all sorts of problems, see <https://github.com/NixOS/nixpkgs/issues/114995>, <https://github.com/NixOS/nixpkgs/issues/118481> and <https://github.com/NixOS/nixos-homepage/issues/672>) and we should be getting rid of them as soon as feasible.

In that sense the RFC solves a problem that would be disappear anyway. Since the new Nix CLI would introduce a replacement for`nix-env` which hopefully doesn't exhibit these flaws.

On the other hand having a solution for the meantime could be interesting, since the new `nix profile` subcommand seems to heavily rely on flakes which means it can't be a stable replacement for `nix-env` in Nix 2.4. This is just speculation though, I'm not sure what @edolstra has planned here.

Reply:
But the BDFL said there is no flaw in parseDrvName. It just splits name and version from a given string, returning a `{name, version}` set.

I don't think the examples given above point to a flaw in it.